### DISCUSSION TITLE
Grammar-based evolution

### DISCUSSION THREAD

**Discussion by hrstoyanov** [jenetics/jenetics#812] [Ideas]
H @jenetics ,
I took a look at the [JENETCS-767 PR](https://github.com/jenetics/jenetics/pull/772) - the code is very professional, using the latest Java 17 features! I do not understand the entire architecture yet, but have some observations/questions I can start with:
- The RegressionExample shows a way to create a program tree off a int array. Is there a way to use ```java.util.BitSet``` instead? Or do you use ```BitGene/BitChromosome/BitArray```?
-  Do not quite understand the  purpose of ```Sentence```? Is this related with recursive rules and max recursion depth?
-  I like the concept of CODEC! The important thing is that any  sequence of genes always produces a valid Tree under a given Grammar, which allows you to use a standard GA algorithm.
- As you implement this, pls pay attention to memory efficiency. Some GAs run huge population where every object counts.
- What is Purdom?
- In ```Parser``` why would you do this:
 ```java
 abstract class Parser<T extends Token> {
	private final Tokenizer<? extends T> _tokenizer;
  ...
 ```
instead of:
```java
abstract class Parser<T extends Token> {
	private final Tokenizer<T> _tokenizer;
	...
```

- Are you going to refactor Jenetics into Java modules? I think I saw a PR for this? That would be awesome!
- You may want to let the user decide if they want defensive copies of the collections in Cfg, for example:
```java
new Cfg(List.copyOf(nonTerminal), List.copyOf(terminals), List.copyOf(rules),start);
```
vs
```java
new Cfg(nonTerminal, terminals, rules, start);
```

As I mentioned, I only spent limited amount of time looking at JENETCS-767, but I am very exited with the code quality and the rest of Jenetics. I will spend some ,more time playing with it next few weeks, hopefulyy giving you better feedback and contributions. Maybe be we can incorporate @ericmedvet more advanced ideas and contributions, as this starts to look more promising than JGEA.

**Comment by jenetics** [comment]
Hi, thank you for the feedback.

* The `Codons` class can be created from various sources, till now I concentrated on chromosomes as source, but it is no problem to use also a `BitSet`.
* The `Sentence` class only contains some factory methods for creating the codecs and other things concerning sentence creation.
* My experiments shows that I can't guarantee that the sentence generation terminates, in a reasonable time. This is way I introduced a sentence limit. The following diagrams shows quick results of these experiments. I limited the sentence length to 200. A sentence length of zero means that the sentence generation was interrupted, because the maximal length was exceeded.
![bit_chromosome_sentences](https://user-images.githubusercontent.com/1509203/145875930-f561a781-2262-4b9a-9e45-9d0d625ea670.png)
![integer_chromosome_sentences](https://user-images.githubusercontent.com/1509203/145875942-10b33448-dc2f-44e8-8179-deacddab44d4.png)
![random_sentences](https://user-images.githubusercontent.com/1509203/145875953-527f2af0-111e-45d3-bbee-3c2b705287e2.png)
* I explored some sentence generation algorithms and Purdom's algorithm is one of them. https://ieeexplore.ieee.org/document/5254270
* Yes, `Tokenizer<T> _tokenizer;` would be better.
* I stopped the migration to modules, because of the (missing) Gradle support. I had to add `SuppressWarnings` annotation on the module declaration and that doesn't felt right.
* Efficiency is of course a thing, but I first tried to find the right model.
* I designed the `Cfg` class as immutable class and the rest of the code is relying on this. I don't think that this will be a performance problem, since the Cfg is created only once for the whole GE run.

The code is still in an early stage, but the first implementations looks also very promising to me. However, I plan to release a first version in the _incubation_ module, to have time to settle the design and API.

**Comment by jenetics** [comment]
My adapted CFG to Genotype mapping will look like this. Every rule is encoded with the correct range. No need for doing a modulo operation when getting the next index from the codons, with all the drawbacks on the uneven index representations.

![adapted_cfg_genotpy_mapping](https://user-images.githubusercontent.com/1509203/145885042-8e15e940-eefd-401e-9be3-3b1f424693e8.png)

**Comment by hrstoyanov** [comment]
Thanks, will review further!

**Comment by jenetics** [comment]
The GE implementation is in its final state. I moved the classes into the `io.jenetics.ext` module and `io.jenetics.ext.grammar` package. The `Cfg.Terminal` class can now  an additional generic value. This will be helpful when _sentences_, generated by the grammar has to be processed further. A possible usage example is implemented in the [CreationAndParsingTest](https://github.com/jenetics/jenetics/blob/issues/JENETICS-815-grammatical_evolution/jenetics.ext/src/test/java/io/jenetics/ext/grammar/CreationAndParsingTest.java) class. It shows also the helper classes/methods, which makes the usage of GE easier.

So, if you want @hrstoyanov, have a look at the implementation: https://github.com/jenetics/jenetics/tree/issues/JENETICS-815-grammatical_evolution/jenetics.ext/src/main/java/io/jenetics/ext/grammar

**Comment by hrstoyanov** [comment]
I will take a look, thanks.

/Hristo Stoyanov
